caddytls: Add `dns_ttl` config, improve Caddyfile `tls` options

ihmels opened the pull request:
My DNS provider (INWX) requires the TTL of a TXT record to be at least 300 seconds. Currently there is no way to configure this in a Caddyfile, if I see it correctly.

In addition, the `propagation_delay` and `propagation_timeout` options have also been added to the `tls` directive.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/caddyserver/caddy?pullRequest=5287) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/caddyserver/caddy?pullRequest=5287) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/caddyserver/caddy?pullRequest=5287) it.</sub>

mholt reviewed (approved):
Interesting; LGTM, thanks for the contribution and the test!

francislavoie commented:
I think this option should also be added to the `tls` directive as well, i.e. in `caddyconfig/httpcaddyfile/builtins.go` so that it can be configured without `issuer`. This is important because if you configure `issuer acme` only, then you disable the default issuer fallback to `zerossl`.

There's a few other properties that should probably get that treatment as well, such as `propagation_timeout` and `propagation_delay`, I just forgot to do those when I implemented those (if you'd like to make that change as well in this PR, that'd be cool :+1:)

mholt commented:
@francislavoie DNS TTL is specific to the ACME challenge though. I think placing it in the `tls` directive is out of scope. (It's already the case that if you want to customize the ACME issuers you need to manually specify the issuers you want.)

francislavoie commented:
That's not true @mholt, some of the options in top-level `tls` propagate down to both issuers, such as `dns`, `dns_challenge_override_domain` and `resolvers`. This is important so that it's easy to set these options both both issuers at once from the Caddyfile, without breaking the default issuer fallback behaviour.

ihmels commented:
I've now added the "dns_ttl" option to the TLS directive and also added "propagation_delay" and "propagation_timeout".

I've also added missing options in the godoc.

francislavoie commented on caddyconfig/httpcaddyfile/builtins.go:
```diff
@@ -363,6 +370,75 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
+				if acmeIssuer.Challenges.DNS == nil {
+					acmeIssuer.Challenges.DNS = new(caddytls.DNSChallengeConfig)
+				}
+				acmeIssuer.Challenges.DNS.TTL = caddy.Duration(delay)
```
I think this is setting the wrong field (propagation delay)

ihmels replied:
Oops, you're right. I've fixed it. :grin:

francislavoie commented on caddyconfig/httpcaddyfile/builtins.go:
```diff
@@ -363,6 +370,75 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
+				if acmeIssuer.Challenges.DNS == nil {
+					acmeIssuer.Challenges.DNS = new(caddytls.DNSChallengeConfig)
+				}
+				acmeIssuer.Challenges.DNS.TTL = caddy.Duration(timeout)
```
Same here (propagation timeout)

francislavoie reviewed (approved):
Great, thank you!

I guess we could use a test for the top-level `tls` config + propagation config (or add it to an existing test) for bonus points, but this meets the minimum bar :grin:

mholt reviewed (approved):
Ah yeah, I guess @francislavoie is right as usual :)

Thank you for the discussion and the contribution. Merge away!
